AMBARI-24742. Encrypting/decrypting PASSWORD type properties when inserting them into the DB/using them

smolnar82 opened the pull request:
## What changes were proposed in this pull request?

Sensitive service configuration values should be encrypted in the Ambari server DB, if enabled.

Sensitive service configuration values are defined by a service's configuration metadata. Properties are defined in XML files under the service's definition directory and contain attributes that Ambari may use to determine whether they should be encrypted or not.

Currently, Ambari uses the property-type attribute to determine the type of property. If the value of this attribute is `PASSWORD`, than the value is considered sensitive and should be encrypted.

The Ambari server should encrypt sensitive configuration values if the following has been met:
1. A master key has been setup using the `ambari-server setup-security` CLI (using option #2 - Encrypt passwords stored in ambari.properties file)
2. The Ambari server configuration property named `server.security.encrypt_sensitive_data` is set to `true`

## How was this patch tested?

In addition to unit testing I executed several E2E tests:
After creating a new `ssl-server` configuration:

<img width="1600" alt="screen shot 2018-10-14 at 1 35 13 am" src="https://user-images.githubusercontent.com/34065904/46942301-51f7b100-d06d-11e8-8014-d89497c4b369.png">

Checking my changes using Ambari API (modified the code a little bit temporarily to avoid secret reference replacement):
<img width="1093" alt="screen shot 2018-10-14 at 1 34 53 am" src="https://user-images.githubusercontent.com/34065904/46942283-4ad0a300-d06d-11e8-9ead-9ee55a46090e.png">

smolnar82 committed:
AMBARI-24742. Encrypting/decrypting PASSWORD type properties when inserting them into the DB/using them

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4271/
Test FAILed.
Test FAILured.

smolnar82 commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4273/
Test PASSed.

rlevas commented on ambari-server/src/main/java/org/apache/ambari/server/security/encryption/PasswordPropertiesEncryptor.java:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  @Override
+  public Map<String, String> encryptSensitiveData(Map<String, String> properties, Object... additionaKeys) {
```
Can you document what the additional keys are for this implementation?

smolnar82 replied:
will do

rlevas commented on ambari-server/src/main/java/org/apache/ambari/server/security/encryption/PasswordPropertiesEncryptor.java:
```diff
@@ -0,0 +1,121 @@
+        final Cluster cluster = (Cluster) additionaKeys[0];
+        final String configType = (String) additionaKeys[1];
+        final Set<String> passwordProperties = getPasswordProperties(cluster, configType);
+        if (passwordProperties != null && !passwordProperties.isEmpty()) {
```
I think we tend to use `CollectionUtils.isNotEmpty(passwordProperties)` instead of the long form here.

smolnar82 replied:
nice catch; I'll change it

rlevas commented on ambari-server/src/main/java/org/apache/ambari/server/security/encryption/PasswordPropertiesEncryptor.java:
```diff
@@ -0,0 +1,121 @@
+  }
+
+  private Set<String> getPasswordProperties(Cluster cluster, String configType) throws AmbariException {
+    final long clusterId = cluster.getClusterId();
```
How will a stack upgrade affect this?

smolnar82 replied:
nice catch; I'll handle that situation like this: subscribing to org.apache.ambari.server.events.StackUpgradeFinishEvent and clean the map when this happened.

Any objection?

rlevas replied:
No objection here.  Thanks.

rlevas replied:
During a stack upgrade, new properties are added and possibly properties change types.   How will this cache behave?

smolnar82 replied:
Upon stack upgrade finish the cache will be purged. So that proceeding encryption actions will re-build it.
See the latest patch I submitted (grep for `onStackUpgradeFinished`)

smolnar82 committed:
AMBARI-24742. Handling the situation when stack is upgraded; documented additional information we need for encryption and using 3rd party to check if password property set of a config type is netiher null nor empty

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4277/
Test FAILed.
Test FAILured.

smolnar82 commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4278/
Test FAILed.
Test FAILured.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4279/
Test FAILed.
Test FAILured.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4280/
Test PASSed.

zeroflag commented on ambari-server/src/main/java/org/apache/ambari/server/security/encryption/PasswordPropertiesEncryptor.java:
```diff
@@ -20,45 +20,61 @@
+   */
   @Override
-  public Map<String, String> encryptSensitiveData(Map<String, String> properties, Object... additionaKeys) {
+  public Map<String, String> encryptSensitiveData(Map<String, String> properties, Object... additionaInfo) {
```
Any reason why we pass these arguments as a Object[] ? I see it comes from the interface, but the caller still needs to know which is the underlying implementation in use to fill these params correctly.

smolnar82 replied:
But the caller should not be down casted to the implementation.

dlysnichenko commented on ambari-server/src/main/java/org/apache/ambari/server/security/encryption/PasswordPropertiesEncryptor.java:
```diff
@@ -0,0 +1,150 @@
+    if (!clusterPasswordProperties.containsKey(clusterId)) {
+      clusterPasswordProperties.put(clusterId, new ConcurrentHashMap<>());
+    }
+    if (!clusterPasswordProperties.get(clusterId).containsKey(configType)) {
```
I think it may be replaced with `java.util.concurrent.ConcurrentHashMap#computeIfAbsent` . As a bonus, we would get atomic check/assigment

smolnar82 replied:
Thanks for this hint (I've to admit I did not know about this option)

dlysnichenko commented on ambari-server/src/main/java/org/apache/ambari/server/security/encryption/PasswordPropertiesEncryptor.java:
```diff
@@ -0,0 +1,150 @@
+
+  @Override
+  public Map<String, String> decryptSensitiveData(Map<String, String> properties, Object... additionalInfo) {
+    final Map<String, String> decryptedProperties = new HashMap<>(properties.size());
```
Are not we replacing here an instance of ConcurrentHashMap with an instance of a plain HashMap (that is unsafe for concurrent use)? I think we access and modify cluster properties from different threads

smolnar82 replied:
Those properties are generated by ConfigFactory after they have been parsed out from the request: https://github.com/apache/ambari/blob/trunk/ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ConfigurationResourceProvider.java#L139
As you can see this is a simple HashMap. Which is ok since we guard the actual Config creation with locks and can not be happening that we have multiple properties map for the same version.

Moreover, those properties are being queried in ConfigImpl.getProperties(), which did simply wrapped the internal properties map in a HashMap in previous versions and does it here too.

smolnar82 committed:
AMBARI-24472. Using a more effective way to populate password properties cache

smolnar82 committed:
AMBARI-24742. Caching properties types by stacks

smolnar82 commented:
Tested the following use cases (sensitive data encryption is turned on):
- updated config before upgrade
- did a stack upgrade (added a new PASSWORD type property to cover that case too)
- updated config after upgrade

Everything went fine.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4329/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4335/
Test PASSed.

smolnar82 commented:
I had to rebase locally and send a new patch to be able to merge with trunk

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4336/
Test FAILed.
Test FAILured.

smolnar82 commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4338/
Test PASSed.

smolnar82 committed:
AMBARI-24742. To make encryptor really generic we implement a Config encryptor

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4340/
Test FAILed.
Test FAILured.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4341/
Test FAILed.
Test FAILured.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4342/
Test FAILed.
Test FAILured.

smolnar82 committed:
AMBARI-24742. Missing decryption when loading data from DB

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4343/
Test FAILed.
Test FAILured.

smolnar82 commented:
Rebased locally with a new change to avoid Swagger issue (unrelated to my changes)

rlevas commented on ambari-server/src/main/java/org/apache/ambari/server/security/encryption/ConfigPropertiesEncryptor.java:
```diff
@@ -0,0 +1,127 @@
+
+  private Set<String> getPasswordProperties(Cluster cluster, StackId stackId, String configType) {
+    final long clusterId = cluster.getClusterId();
+    clusterPasswordProperties.computeIfAbsent(clusterId, v -> new ConcurrentHashMap<>()).computeIfAbsent(stackId, v -> new ConcurrentHashMap<>())
```
Why bother caching using the cluster id at all.  Even if Ambari was managing multiple clusters, the stack definitions all would be the same for the same stack id.

smolnar82 replied:
hm..that's true. I may remove that layer next week.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4345/
Test PASSed.
